Documentation.app/Contents/MacOS/FP concepts.playground/Pages/Rank-N polymorphism.xcplaygroundpage/Contents.swift

```diff
@@ -0,0 +1,155 @@
+
+ ## The basics of closures
+ Swift supports higher-order functions, which means that we can pass a function as an argument to another function or return a function as the result of another function. Furthermore, we can define *closures* which are functions that capture some data from its environment.
+
+ For example, the function `randomWordGenerator` below returns a random word of length `n`. As you can see `n` is a random number that is generated once each time you execute this program. `randomWordGenerator` captures the value of `n`, which means that the strings returned by `randomWordGenerator` will all have the same length during the execution of the program. Each time you call `randomWordGenerator` you will get a different word, but all of them will have the same length. If you re-run the program a second time, `n` will probably get a different value and the length of the words returned by `randomWordGenerator` will be different.
+*/
+let n: Int = .random(in: 0..<10)
+let randomWordGenerator: () -> String = { (0..<n).forEach { _ in .random(in: "a"..."z") } }
+/*:
+ In the example above we only got one `randomWordGenerator` function per program execution: if we want a different `randomWordGenerator` function with a different `n` we need to re-run the program. Let's create a function that returns a different `randomWordGenerator` function each time we call it. This is an example of a higher-order function, since it returns another function.
+ */
+func makeRandomWordGenerator() -> () -> String {
+    let length: Int = .random(in: 0..<10)
+    return { (0..<length).forEach { _ in .random(in: "a"..."z") } }
+}
+/*:
+ The body of `makeRandomWordGenerator` is basically the same code we had before (we renamed `n` to `length` to stress the fact that `makeRandomWordGenerator` doesn't use any variables from outside its body). But now, each time we call `makeRandomWordGenerator` a different `length` is generated, and thus we get a different `randomWordGenerator` function each time.
+
+ We check this by creating 3 different functions with `makeRandomWordGenerator` and generating one word with each. We should see that each function prints words of different length (unless we are a bit unlucky).
+ */
+for i in 0..<3 {
+    let f = makeRandomWordGenerator()
+    print(f())
+}
+/*:
+ An important thing to realise is that we cannot just generate the random `length` value inside the body of the closure returned from `makeRandomWordGenerator`. If we did so, a `randomWordGenerator` would return strings of different length each time we call it, and this is not what we want. We want each `randomWordGenerator` function to always return strings of the same length.
+
+ This little example demonstrates that closures indeed capture values from their environment. Sine we define `length` inside `makeRandomWordGenerator`,
```

Comment:
```suggestion
 This little example demonstrates that closures indeed capture values from their environment. Since we define `length` inside `makeRandomWordGenerator`,
```